### ISSUE TITLE
Eliminate the second type parameter of punctuated::Iter, IterMut, IntoIter

### ISSUE THREAD

**Issue by dtolnay** [dtolnay/syn#333]
Currently if you iterate over the values of a `Punctuated<T, P>` the iterator types are `Iter<'a, T, P>`, `IteraMut<'a, T, P>`, `IntoIter<T, P>`. Instead they should be `Iter<'a, T>`, `IterMut<'a, T>`, `IntoIter<T>`.

Relevant paper: http://www.stroustrup.com/SCARY.pdf

For this to be well defined we will need to replace the `Vec<(T, Option<P>)>` with a `Vec<RawPair>` with a defined layout:

```rust
#[repr(C)]
struct RawPair<T, P> {
    t: T,
    p: Option<P>,
}
```

And then have the various iterators store `mem::size_of::<RawPair<T, P>>()` when constructed and use that as a stride when traversing the contents of the vector.

**Comment by mystor** [comment]
I'm skeptical it's worthwhile to introduce unsafe code to optimize this particular case, but it might be worth it to clean up the public interface. I'm not sure.

**Comment by dtolnay** [comment]
I believe the public API improvement warrants this change.

**Comment by sgrif** [comment]
Even the standard library doesn't go this far. `hash_map::Keys` is parameterized over the value for example. https://doc.rust-lang.org/nightly/std/collections/hash_map/struct.Keys.html

Ultimately the public API is effectively `impl Iterator<Item = T>` either way. I don't think this really affects the public API

**Comment by dtolnay** [comment]
This affects the readability of the signature of [`Fields::iter`](https://docs.rs/syn/0.12.6/syn/enum.Fields.html#method.iter) for example.

I give my permission for Syn to handle this better than the standard library did. If Rust 0.9 had implemented `Keys<'a, K>`, and having read the Stroustrup paper, we would not review it for 1.0 and decide that `Keys<'a, K, V>` would be preferable.

There may be a safe way to implement this by having `syn::punctuated::Iter<T>` store a `Box<Iterator<Item = &T>>` that is implemented by a private `Iter<T, P>`.

**Comment by dtolnay** [comment]
Fixed in 8095c30c073611aebe4151571ce480f8fea1dd75.
